### ISSUE TITLE
Memory leak when use model.fit method with datagenerator

### ISSUE THREAD

**Issue by dishkakrauch** [tensorflow/tensorflow#36114]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
it's increasing ram when I use mode.fit() method
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): installed via pip
- TensorFlow version (use command below): tensorflow-gpu 2.1.0
- Python version: 3.7
- Bazel version (if compiling from source): No
- GCC/Compiler version (if compiling from source): No
- CUDA/cuDNN version: 10.1
- GPU model and memory: GeForce GTX 1060 6GB

**Describe the current behavior**
RAM memory increases after I fit model. I cannot clear it with any garbage collector etc.
**Describe the expected behavior**
RAM shoud not increases at each epoch of training any model in that way.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
I'm using my own image generator with multioutput:

**
class MultiOutputDataGenerator(tf.keras.preprocessing.image.ImageDataGenerator):

    def flow(
        self,
        x,
        y=None,
        batch_size=32,
        shuffle=True,
        sample_weight=None,
        seed=None,
        save_to_dir=None,
        save_prefix='',
        save_format='png',
        subset=None
    ):

        targets = None
        target_lengths = {}
        ordered_outputs = []
        
        for output, target in y.items():
            
            if targets is None:
                
                targets = target
                
            else:
                
                targets = np.concatenate((targets, target), axis=1)
                
            target_lengths[output] = target.shape[1]
            ordered_outputs.append(output)
        
            gc.collect()

        for flowx, flowy in super().flow(x, targets, batch_size=batch_size, shuffle=shuffle):
            
            target_dict = {}
            i = 0
            
            for output in ordered_outputs:
                
                target_length = target_lengths[output]
                target_dict[output] = flowy[:, i: i + target_length]
                i += target_length
                
                gc.collect()

            yield flowx, target_dict
**

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by gadagashwini** [comment]
@dishkakrauch,
Could you post the complete code to reproduce the reported issue. Thanks!

**Comment by dishkakrauch** [comment]
> @dishkakrauch,
> Could you post the complete code to reproduce the reported issue. Thanks!

Sure! No problem.
Here is it:

import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
#import tqdm
import gc
gc.enable()
import os
import sys
import graphviz
os.environ["PATH"] += os.pathsep + 'C:/Program Files (x86)/Graphviz2.38/bin/'

import tensorflow as tf
import sklearn
from sklearn import *

import matplotlib.pylab as plt
import matplotlib.image as mpimg
import seaborn as sns
import cv2
import PIL.Image as Image, PIL.ImageDraw as ImageDraw, PIL.ImageFont as ImageFont

IMG_SIZE = 128
N_CHANNELS = 1

HEIGHT = 137
WIDTH = 236

def resize(df, size, need_progress_bar=False):
    resized = {}
    if need_progress_bar:
        for i in tqdm.tqdm_notebook(range(df.shape[0])):
            image = cv2.resize(df.loc[df.index[i]].values.reshape(137,236),(size,size))
            resized[df.index[i]] = image.reshape(-1)
    else:
        for i in range(df.shape[0]):
            image = cv2.resize(df.loc[df.index[i]].values.reshape(137,236),(size,size))
            resized[df.index[i]] = image.reshape(-1)
    resized = pd.DataFrame(resized).T
    return resized

def get_dummies(df):
    cols = []
    for col in df:
        cols.append(pd.get_dummies(df[col].astype(str)))
    return pd.concat(cols, axis=1)

class MultiOutputDataGenerator(tf.keras.preprocessing.image.ImageDataGenerator):

    def flow(
        self,
        x,
        y=None,
        batch_size=32,
        shuffle=True,
        sample_weight=None,
        seed=None,
        save_to_dir=None,
        save_prefix='',
        save_format='png',
        subset=None
    ):

        targets = None
        target_lengths = {}
        ordered_outputs = []
        
        for output, target in y.items():
            
            if targets is None:
                
                targets = target
                
            else:
                
                targets = np.concatenate((targets, target), axis=1)
                
            target_lengths[output] = target.shape[1]
            ordered_outputs.append(output)
        
            gc.collect()

        for flowx, flowy in super().flow(x, targets, batch_size=batch_size, shuffle=shuffle):
            
            target_dict = {}
            i = 0
            
            for output in ordered_outputs:
                
                target_length = target_lengths[output]
                target_dict[output] = flowy[:, i: i + target_length]
                i += target_length
                
                gc.collect()

            yield flowx, target_dict

train_df = pd.read_csv(path + 'train.csv')
df_0 = pd.read_feather(path + train_image_data_0.fth')

inputs = tf.keras.layers.Input(
    shape=(IMG_SIZE, IMG_SIZE, 1)
)

model = tf.keras.layers.Conv2D(
    filters=32,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding='SAME',
    activation=tf.keras.activations.relu,
    input_shape=(IMG_SIZE, IMG_SIZE, 1)
)(inputs)

model = tf.keras.layers.Conv2D(
    filters=32,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding='SAME',
    activation=tf.keras.activations.relu,
    input_shape=(IMG_SIZE, IMG_SIZE, 1)
)(model)

model = tf.keras.layers.BatchNormalization(
    momentum=.15
)(model)

model = tf.keras.layers.MaxPool2D(
    pool_size=(2, 2),
    strides=None
)(model)

model = tf.keras.layers.Dropout(
    rate=.3
)(model)

model = tf.keras.layers.Conv2D(
    filters=64,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding='SAME',
    activation=tf.keras.activations.relu,
    input_shape=(IMG_SIZE, IMG_SIZE, 1)
)(model)

model = tf.keras.layers.Conv2D(
    filters=64,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding='SAME',
    activation=tf.keras.activations.relu,
    input_shape=(IMG_SIZE, IMG_SIZE, 1)
)(model)

model = tf.keras.layers.BatchNormalization(
    momentum=.15
)(model)

model = tf.keras.layers.MaxPool2D(
    pool_size=(2, 2),
    strides=None
)(model)

model = tf.keras.layers.Dropout(
    rate=.3
)(model)

model = tf.keras.layers.Conv2D(
    filters=128,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding='SAME',
    activation=tf.keras.activations.relu,
    input_shape=(IMG_SIZE, IMG_SIZE, 1)
)(model)

model = tf.keras.layers.Conv2D(
    filters=128,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding='SAME',
    activation=tf.keras.activations.relu,
    input_shape=(IMG_SIZE, IMG_SIZE, 1)
)(model)

model = tf.keras.layers.BatchNormalization(
    momentum=.15
)(model)

model = tf.keras.layers.MaxPool2D(
    pool_size=(2, 2),
    strides=None
)(model)

model = tf.keras.layers.Dropout(
    rate=.3
)(model)

model = tf.keras.layers.Flatten()(model)


model = tf.keras.layers.Dense(
    units=1024,
    activation=tf.keras.activations.relu
)(model)

model = tf.keras.layers.Dropout(
    rate=0.3
)(model)

dense = tf.keras.layers.Dense(
    units=512,
    activation = tf.keras.activations.relu
)(model)

head_root = tf.keras.layers.Dense(
    units=168,
    activation=tf.keras.activations.softmax,
    name='root'
)(dense)

head_vowel = tf.keras.layers.Dense(
    units=11,
    activation=tf.keras.activations.softmax,
    name='vowel'
)(dense)

head_consonant = tf.keras.layers.Dense(
    units=7,
    activation=tf.keras.activations.softmax,
    name='consonant'
)(dense)

model = tf.keras.Model(
    inputs=inputs,
    outputs=[head_root, head_vowel, head_consonant]
)

model.compile(
    optimizer=tf.keras.optimizers.Adam(),
    loss=tf.keras.losses.categorical_crossentropy,
    metrics=['accuracy']
)

train_df = train_df.drop(['grapheme'], axis=1, inplace=False)

train_df[['grapheme_root', 'vowel_diacritic', 'consonant_diacritic']] = \
train_df[['grapheme_root', 'vowel_diacritic', 'consonant_diacritic']].astype('uint8')

X_train = pd.merge(
    df_0,
    train_df,
    on='image_id'
).drop(
    labels=['image_id'],
    axis=1
)

Y_train_root = pd.get_dummies(X_train['grapheme_root']).values
Y_train_vowel = pd.get_dummies(X_train['vowel_diacritic']).values
Y_train_consonant = pd.get_dummies(X_train['consonant_diacritic']).values

X_train = X_train.drop(
    labels=['grapheme_root', 'vowel_diacritic', 'consonant_diacritic'],
    axis=1
)
X_train = resize(X_train, size=IMG_SIZE)/255

X_train = X_train.values.reshape(-1, IMG_SIZE, IMG_SIZE, N_CHANNELS)

x_train, x_test, y_train_root, y_test_root, y_train_vowel, y_test_vowel, y_train_consonant, y_test_consonant = \
sklearn.model_selection.train_test_split(
    X_train,
    Y_train_root,
    Y_train_vowel,
    Y_train_consonant,
    test_size=.08,
    random_state=0
)

del X_train
del Y_train_root
del Y_train_vowel
del Y_train_consonant

gc.collect()

datagen = MultiOutputDataGenerator(
    featurewise_center=False,  # set input mean to 0 over the dataset
    samplewise_center=False,  # set each sample mean to 0
    featurewise_std_normalization=False,  # divide inputs by std of the dataset
    samplewise_std_normalization=False,  # divide each input by its std
    zca_whitening=False,  # apply ZCA whitening
    rotation_range=8,  # randomly rotate images in the range (degrees, 0 to 180)
    zoom_range=.15, # Randomly zoom image 
    width_shift_range=.15,  # randomly shift images horizontally (fraction of total width)
    height_shift_range=.15,  # randomly shift images vertically (fraction of total height)
    horizontal_flip=False,  # randomly flip images
    vertical_flip=False  # randomly flip images
)

datagen.fit(x_train)

from memory_profiler import profile

%load_ext memory_profiler

def main():
    
    model.fit(
        datagen.flow(
            x_train,
            {
                'root': y_train_root,
                'vowel': y_train_vowel,
                'consonant': y_train_consonant
            },
            batch_size=batch_size
        ),
        epochs=1,
        validation_data=(x_test, [y_test_root, y_test_vowel, y_test_consonant]),  
        steps_per_epoch=x_train.shape[0] // batch_size, 
        callbacks=[
            callback_lr,
            callback_es
        ]
    )
    
    gc.collect()

# here is where profiler saying that fitiing model increases ram usage
%memit main()

**Comment by dishkakrauch** [comment]
@gadagashwini any suggestions?

**Comment by gadagashwini** [comment]
@dishkakrauch, I tried replicating the issue on colab but getting different error. Please take a look at colab [gist](https://colab.research.google.com/gist/gadagashwini/ec233809cd9bc7d15a0d3f98b459e8c1/untitled22.ipynb) and provide more information to reproduce the issue. Thanks!

**Comment by dishkakrauch** [comment]
> @dishkakrauch, I tried replicating the issue on colab but getting different error. Please take a look at colab [gist](https://colab.research.google.com/gist/gadagashwini/ec233809cd9bc7d15a0d3f98b459e8c1/untitled22.ipynb) and provide more information to reproduce the issue. Thanks!

It's cause you don't have data for training. How can I share it with you?

**Comment by gadagashwini** [comment]
@dishkakrauch, Can you provide data as zip as file here. Thanks

**Comment by dishkakrauch** [comment]
> @dishkakrauch, Can you provide data as zip file here. Thanks

It's more than limit 10MB. Can I share google drive link here?
https://drive.google.com/file/d/1ggk622mLwkpWHS04Q8lq1ITIKZwSPOPh/view?usp=sharing

**Comment by dishkakrauch** [comment]
@gadagashwini got any updates?

**Comment by gadagashwini** [comment]
@dishkakrauch, Sorry for delayed response. I tried replicating the issue with your dataset but received different error, can you take a look at [gist](https://colab.research.google.com/gist/gadagashwini/8eb404091c851c12b97ce27b674b8e65/untitled22.ipynb#scrollTo=kvvxvhrGpw9V) and make changes in gist. Thanks!

**Comment by dishkakrauch** [comment]
> @dishkakrauch, Sorry for delayed response. I tried replicating the issue with your dataset but received different error, can you take a look at [gist](https://colab.research.google.com/gist/gadagashwini/8eb404091c851c12b97ce27b674b8e65/untitled22.ipynb#scrollTo=kvvxvhrGpw9V) and make changes in gist. Thanks!

I'm really surprised that it's a big problem to load file and try to repeat all the following steps to my erorr. Can't you load file with feather-format package insted pandas?

```
pip install feather-format
import feather
df_0 = feather.read_dataframe(path)
```

**Comment by dishkakrauch** [comment]
Solved by:
import tensorflow as tf
tf.keras.backend.clear_session()
Thanks to @taborda11
